More type-hints - Batch 003

Quix0r opened the pull request:
A new batch of changes of type-hints, added missing documentation, fixing bad method calls.

Quix0r committed:
Changes:
- added more type-hints

Quix0r committed:
Changes:
- added type-hints
- added missing documentation

Quix0r committed:
Changes:
- added some documentation
- added type-hints

Quix0r committed:
Changes:
- added type-hints
- added some missing documentation

Quix0r committed:
Ops, wrong type-hint

Quix0r committed:
Changed:
- cannot have type-hints :-(

Quix0r committed:
Both declarations must be the same

Quix0r committed:
Changes:
- some methods now need to return bool to be compatible
- added some missing type-hints

Quix0r committed:
Changed:
- DBA::exists() should only be used for checking if records exists.
- if you want to check if a table exists, please ALWAYS use
  DBStructure::existsTable() instead

Quix0r committed:
Fixes:
- Strings::isHex() should not be misused for checking on NULL

Quix0r commented:
Some important changes:
- `DBA::exists()` should only be used for checking if records exists.
- if you want to check if a table exists, please ALWAYS use `DBStructure::existsTable()` instead
- `Strings::isHex()` should not be misused for checking on `NULL`

Quix0r commented:
If you want to have a `DBA::existsTable()`, just add it. :-)

MrPetovan commented on src/Database/Database.php:
```diff
@@ -1504,6 +1526,7 @@ public function select($table, array $fields = [], array $condition = [], array
 	 */
 	public function count($table, array $condition = [], array $params = [])
 	{
+		// @TODO Can we dump this to have ": int" as returned type-hint?
```
Yes. The error state is taken care of in line 1550 anyway. A better solution would be to throw an `InvalidArgumentException` which should never trigger since it'd be a developer error.

Quix0r replied:
I would also prefer an `IAE` here instead of returned `false`.

Quix0r replied:
But I guess this process need to be done slowly.

MrPetovan commented on tests/src/Util/StringsTest.php:
```diff
@@ -113,22 +113,18 @@ public function dataIsHex()
 	 * @dataProvider dataIsHex
 	 */
-	public function testIsHex(string $input = null, bool $valid = false)
+	public function testIsHex(string $input = '', bool $valid = false)
```
You can remove the default value now that we don't test null anymore.

Quix0r committed:
Changes:
- dumped default value for $input
- added unknown 'platform' which prevents an "Undefined index: platform in
  /var/www/.../src/Model/GServer.php on line 940" error

Quix0r committed:
Changes:
- fixed a null value handled over to Friendica\Model\APContact::getByURL()
- added missing type-hints

Quix0r committed:
Changes:
- let's start throwing exceptions on e.g. invalid arguments instead of returning
  FALSE

Quix0r committed:
Changes:
- null was 2nd argument's value before, an empty string is basically the same here

MrPetovan commented on src/Protocol/ActivityPub/Receiver.php:
```diff
@@ -98,7 +97,7 @@ public static function processInbox($body, $header, int $uid)
 
 		$actor = JsonLD::fetchElement($ldactivity, 'as:actor', '@id');
```
`$actor` is used a few times, might as well do the null-coalescing earlier.

```suggestion
		$actor = JsonLD::fetchElement($ldactivity, 'as:actor', '@id') ?? '';
```

Quix0r replied:
Okay, that proposed change makes sense!

MrPetovan commented on src/Protocol/ActivityPub/Receiver.php:
```diff
@@ -98,7 +97,7 @@ public static function processInbox($body, $header, int $uid)
 		$actor = JsonLD::fetchElement($ldactivity, 'as:actor', '@id');
 
-		$apcontact = APContact::getByURL($actor);
+		$apcontact = APContact::getByURL($actor ?? '');
```
```suggestion
		$apcontact = APContact::getByURL($actor);
```

Quix0r committed:
Changes:
- as @MrPetovan pointed out, $actor can be NULL earlier and used later as NULL
- added some missing type-hints
- added missing documentation
- the added @TODO points out to avoid true|false|null for a boolean

Quix0r committed:
Added documentation

Quix0r committed:
Changes:
- renamed ItemArrayFromMail() to getItemArrayFromMail() to follow naming-convetion
- added missing type-hints
- added missing documentation

Quix0r commented:
Do you want me to rebase this branch after you have reviewed it?

Quix0r committed:
Changes:
- added type-hints
- added some documentation
- marked some generic methods to be moved to Util\Strings class instead

MrPetovan commented on src/Protocol/ActivityPub/Transmitter.php:
```diff
@@ -2223,23 +2224,36 @@ public static function sendContactUndo($target, $cid, $uid)
+			],
 			'instrument' => self::getService(),
-			'to' => [$profile['url']]];
+			'to' => [$profile['url']]
```
Adding a trailing comma after each array element, even the last one, prevents possible syntax errors when adding more elements and limit the number of changed lines in git diffs.
```suggestion
			'to' => [$profile['url']],
```

Quix0r replied:
Should be fixed by now.

Quix0r committed:
Fixed indenting

Quix0r committed:
Fixed:
- prevent NULL from being handled over to XML::escape()
- still I wonder that an object can be? It is a string-only accepting method

MrPetovan commented on src/Util/XML.php:
```diff
@@ -56,7 +56,7 @@ public static function fromArray(array $array, &$xml, bool $remove_header = fals
 					self::fromArray($value, $root, $remove_header, $namespaces, false);
 				} else {
-					$root = new SimpleXMLElement('<' . $key . '>'.self::escape($value).'</' . $key . '>');
+					$root = new SimpleXMLElement('<' . $key . '>' . self::escape($value ?? '') . '</' . $key . '>');
```
With self-closing tags, the inner value would be `null`.

Quix0r replied:
`null` will normally result in an empty string: `$str = (string) null; echo gettype($str) . '/' . strlen($str);`

MrPetovan replied:
Yes but it is explicitly forbidden with a type-hinted string parameter without a `null` default value, as you've discovered time and again.
